File: examples/streaming_chat.py
#!/usr/bin/env python3
"""
Streaming Chat Example for RawrXD

This example demonstrates how to use streaming responses
from the RawrXD API in Python.

Requirements:
    pip install requests

Usage:
    python streaming_chat.py
"""

import requests
import json
import sys

# Configuration
API_BASE = "http://localhost:8080/v1"
MODEL = "llama-2-7b"


def stream_chat_completion(messages, temperature=0.7):
    """
    Stream chat completion responses from RawrXD.
    
    Args:
        messages: List of message dictionaries
        temperature: Sampling temperature
    
    Yields:
        Text chunks as they arrive
    """
    response = requests.post(
        f"{API_BASE}/chat/completions",
        json={
            "model": MODEL,
            "messages": messages,
            "temperature": temperature,
            "stream": True
        },
        stream=True
    )
    
    if response.status_code != 200:
        raise Exception(f"API Error: {response.status_code}")
    
    for line in response.iter_lines():
        if line:
            line = line.decode('utf-8')
            if line.startswith('data: '):
                data = line[6:]
                if data == '[DONE]':
                    return
                
                try:
                    parsed = json.loads(data)
                    content = parsed.get('choices', [{}])[0].get('delta', {}).get('content', '')
                    if content:
                        yield content
                except json.JSONDecodeError:
                    pass


def main():
    print("RawrXD Streaming Chat Example")
    print("=" * 50)
    
    # Example 1: Simple streaming
    print("\n1. Simple Streaming Chat:")
    messages = [{"role": "user", "content": "Count from 1 to 5"}]
    
    print("User: Count from 1 to 5")
    print("Assistant: ", end='', flush=True)
    
    for chunk in stream_chat_completion(messages):
        print(chunk, end='', flush=True)
    print("\n")
    
    # Example 2: Story generation
    print("\n2. Story Generation:")
    messages = [{"role": "user", "content": "Write a one-sentence story about AI."}]
    
    print("User: Write a one-sentence story about AI.")
    print("Assistant: ", end='', flush=True)
    
    for chunk in stream_chat_completion(messages):
        print(chunk, end='', flush=True)
    print("\n")
    
    # Example 3: Interactive chat
    print("\n3. Interactive Chat (type 'quit' to exit):")
    conversation = []
    
    while True:
        user_input = input("\nYou: ").strip()
        if user_input.lower() == 'quit':
            break
        
        conversation.append({"role": "user", "content": user_input})
        
        print("Assistant: ", end='', flush=True)
        response_text = ""
        
        for chunk in stream_chat_completion(conversation):
            print(chunk, end='', flush=True)
            response_text += chunk
        
        print()  # New line after response
        conversation.append({"role": "assistant", "content": response_text})


if __name__ == "__main__":
    main()

File: examples/streaming_chat.js
/**
 * Streaming Chat Example for RawrXD
 * 
 * This example demonstrates how to use streaming responses
 * from the RawrXD API in JavaScript.
 * 
 * Usage:
 *   node streaming_chat.js
 */

const API_BASE = 'http://localhost:8080/v1';
const MODEL = 'llama-2-7b';

/**
 * Make a streaming chat completion request
 */
async function streamChatCompletion(messages, onChunk) {
    const response = await fetch(`${API_BASE}/chat/completions`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            model: MODEL,
            messages: messages,
            stream: true
        })
    });

    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();

    while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        const chunk = decoder.decode(value);
        const lines = chunk.split('\n');

        for (const line of lines) {
            if (line.startsWith('data: ')) {
                const data = line.slice(6);
                if (data === '[DONE]') return;

                try {
                    const parsed = JSON.parse(data);
                    const content = parsed.choices[0]?.delta?.content;
                    if (content) {
                        onChunk(content);
                    }
                } catch (e) {
                    // Ignore parse errors
                }
            }
        }
    }
}

/**
 * Example usage
 */
async function main() {
    console.log('RawrXD Streaming Chat Example');
    console.log('=' .repeat(50));

    // Example 1: Simple streaming chat
    console.log('\n1. Simple Streaming Chat:\n');
    
    const messages = [
        { role: 'user', content: 'Count from 1 to 5' }
    ];

    process.stdout.write('Assistant: ');
    await streamChatCompletion(messages, (chunk) => {
        process.stdout.write(chunk);
    });
    console.log('\n');

    // Example 2: Story generation with streaming
    console.log('\n2. Story Generation:\n');
    
    const storyMessages = [
        { role: 'user', content: 'Write a one-sentence story about a robot.' }
    ];

    process.stdout.write('Assistant: ');
    await streamChatCompletion(storyMessages, (chunk) => {
        process.stdout.write(chunk);
    });
    console.log('\n');

    // Example 3: Code generation with streaming
    console.log('\n3. Code Generation:\n');
    
    const codeMessages = [
        { role: 'system', content: 'You are a coding assistant.' },
        { role: 'user', content: 'Write a JavaScript function to reverse a string.' }
    ];

    process.stdout.write('Assistant:\n');
    await streamChatCompletion(codeMessages, (chunk) => {
        process.stdout.write(chunk);
    });
    console.log('\n');
}

main().catch(console.error);
